### ISSUE TITLE
beachball plotting issues

### ISSUE THREAD

**Issue by celsoa** [obspy/obspy#2388]
* conda 4.5.11
* obspy 1.1.1
* python 3.7.1

Hello, the obspy beachball routines seem to plot some beachballs incorrectly. This may be related to numerical instabilities and large numbers mentioned before (#2114), but seems that small numbers can also cause issues. The following examples should all plot solid beachballs (+ISO)

command line
```
obspy-mopad plot -i USE 0.5774,0.5774,0.5774,0,0,0  # solid beachball
obspy-mopad plot -i USE 0.5774,0.5773,0.5774,0,0,0  # 0.0001 difference in one element
```

script
```
from obspy.imaging.beachball import beachball as bb1
from obspy.imaging.mopad_wrapper import beachball as bb2

mt1 = [3.258319e+16,  3.258250e+16,  3.258180e+16,  0.000000e+00,  0.000000e+00,  0.000000e+00]
mt2 = [3.258319    ,  3.258250    ,  3.258180    ,  0           ,  0           ,  0]

fig = bb1(mt1)  # strange beachball
fig = bb1(mt2)  # OK
fig = bb2(mt1)  # strange beachball
fig = bb2(mt2)  # strange beachball
```

right beachball
![rightbb](https://user-images.githubusercontent.com/5817670/56826953-bfc33780-685d-11e9-9603-9ff9f9a6dcdb.png)

strange beachball
![wrongbb](https://user-images.githubusercontent.com/5817670/56826968-c9e53600-685d-11e9-8aae-6295fb120d6f.png)

For reference the following dataset contains a broad range of mechanisms
https://github.com/carltape/compearth/blob/master/momenttensor/plot/gmt/beachball_plots/beachballs_ipts1_iref1_lune_psmeca
`[X, Y, depth, mrr, mtt, mff, mrt, mrf, mtf, exp]`

On an (X,Y) grid they should plot as the following figure
![bbcollection2](https://user-images.githubusercontent.com/5817670/56745788-4bb06300-677b-11e9-97c5-86a011cca7a5.jpg)

**Comment by carltape** [comment]
As a sanity check, each of the two example moment tensors (mt1, mt2) has three eigenvalues that are all positive. So the beachballs should be all solid. From other tests, it appears that the plotting issues extend beyond the all-solid beachball cases.

**Comment by megies** [comment]
Unfortunately, we don't have anybody with a good overview of the existing beachball routines in obspy, I think. It was ported/integrated from [mopad](https://github.com/geophysics/MoPaD) but that is orphaned for several years as far as I can see.

So, I don't really know how to proceed with improving/fixing the current status. Are there other Python codes for beachballs plotting etc that we are not aware of?

**Comment by avuan** [comment]
see https://djpugh.github.io/MTfit/mtplot.html#beachball-plot

**Comment by celsoa** [comment]
@megies the pyrocko beachball routines plot all the beachballs correctly (thanks @emolch)
https://pyrocko.org/docs/current/_modules/pyrocko/moment_tensor.html#MomentTensor
![bbpr](https://user-images.githubusercontent.com/5817670/57789536-69009d80-7739-11e9-8f3c-0e8cd50c9ca8.png)

at the moment the same dataset plotted with obspy looks like this
![bbop](https://user-images.githubusercontent.com/5817670/57789541-6f8f1500-7739-11e9-9a52-40d49577d710.png)

ideally there is a single development branch for plotting beachballs, and I'm not sure the best way forward, but maybe through this pyrocko compatibility package?
https://pyrocko.org/docs/current/library/examples/obspy_compat.html

**Comment by megies** [comment]
@celsoa your top-most example is fixed by #2400, at least when using `obspy.imaging.beachball`. Can you please check if this fixes all of your other examples as well?

**Comment by celsoa** [comment]
@megies it works only for two, an all-red and an all-white beachball (the top and bottom beachballs in my grid figure).

**Comment by celsoa** [comment]
on second check, it looks like the `obspy-mopad` wrapper by default does not plot the isotropic parameter. the first two examples with `obspy-mopad` work when specifying flag `--show-isotropic-part`.

**Comment by megies** [comment]
@celsoa are you sure about the different coordinate systems as well? the variable names `mrr, mtt, mff, mrt, mrf, mtf` sound like they might be in a different coordinate system then what is used in obspy maybe..?

**Comment by celsoa** [comment]
@megies yes, the dataset in my first post is in global CMT format (Up, South, East) = (rho, theta, phi). This also looks to be the default for `obspy.imaging.beachball` and `obspy.imaging.mopad_wrapper`

https://docs.obspy.org/packages/autogen/obspy.imaging.beachball.beachball.html#obspy.imaging.beachball.beachball

https://docs.obspy.org/packages/autogen/obspy.imaging.mopad_wrapper.beachball.html#obspy.imaging.mopad_wrapper.beachball

**Comment by ThomasLecocq** [comment]
@megies what is the target ? The pyrocko reference above, or ?

**Comment by megies** [comment]
@ThomasLecocq apparently we should get the image like in first post, assuming that coordinate systems used are the same, which they apparently are, see https://github.com/obspy/obspy/issues/2388#issuecomment-495271829

**Comment by celsoa** [comment]
@ThomasLecocq @megies to summarize the above:

**right** | **wrong**
------|------
(pyrocko, psmeca) | (obspy, mopad_wrapper)
![bbpr](https://user-images.githubusercontent.com/5817670/64866468-c6e7b800-d63b-11e9-8e2e-0613a6e91156.png) | ![bbop](https://user-images.githubusercontent.com/5817670/64866492-cea75c80-d63b-11e9-9205-3bb107c4b431.png)

The moment tensors for these are listed in the following table (GCMT format)
``` [X, Y, depth, mrr, mtt, mff, mrt, mrf, mtf, exp]```
https://github.com/carltape/compearth/blob/master/momenttensor/plot/gmt/beachball_plots/beachballs_ipts1_iref1_lune_psmeca
the same dataset might also be useful to unit test the beachball routines once they're fixed.

**Comment by carltape** [comment]
I wanted to mention that the GMT developments are ongoing [here](https://github.com/GenericMappingTools/gmt/issues/661). GMT 6.0.0 seems to work for 5/6 test sets of moment tensors, but not (yet) the sixth set of lambda2=0 moment tensors. I encourage folks to plot all 6 sets (and more!) for the milestones. Thanks so much for working on this and documenting the issue.

**Comment by megies** [comment]
Closed this by accident actually, but I'll bump this to a future release since nobody is working on this

**Comment by brioglade** [comment]
I also face this problem that the same CMT seismic source parameters from HybridMT with different bechball by obspy-mopad.

**Comment by carltape** [comment]
I wanted to update some paths to files in a repository that were referenced in this issue. In case this issue is still unresolved, these revised links could be helpful. The input text files of psmeca moment tensors are [here](https://github.com/carltape/mtbeach/tree/master/gmt/dfiles).
The main text file in psmeca format is [beachballs_ipts1_iref1_lune_psmeca](https://github.com/carltape/mtbeach/blob/master/gmt/dfiles/beachballs_ipts1_iref1_lune_psmeca)
The minimal plot is [lune_ngeo_iplot2_lplot1_kplot1_min.pdf](https://github.com/carltape/mtbeach/blob/master/gmt/beachball_plots/lune_ngeo_iplot2_lplot1_kplot1_min.pdf)
A larger plot based on more test moment tensors is [lune_ngeo_iplot2_lplot1_kplot1_gmt611.pdf](https://github.com/carltape/mtbeach/blob/master/gmt/beachball_plots/lune_ngeo_iplot2_lplot1_kplot1_gmt611.pdf)
I did not test these in the latest obspy plotting tools. Thanks.

**Comment by ammcpherson** [comment]
Hello,

I have updated the built-in ObsPy beachball module to show the isotropic component of a moment tensor if desired. These updates also include bug-fixes in the plot_mt function of obspy/imagining/beachball.py such as not fully accounting for large jumps in azimuth and takeoff angles that need a new segment to be built, not completely testing for the dominant eigenvalue, and not always using the correct cfill for pressure and tension quadrants.

I have not worked on the mopad wrapper, so this issue still persists there.

Below I have included some figures from the built-in beachball module that can be compared to the beachballs shown [here](https://github.com/carltape/mtbeach/blob/master/gmt/beachball_plots/lune_ngeo_iplot2_lplot1_kplot1_gmt611.pdf).

<img width="2000" height="1000" alt="Image" src="https://github.com/user-attachments/assets/d32d1955-9259-4e53-a340-2fdf8b4b44d5" />

I would like to contribute this back into the main ObsPy repo. I have forked it and created a new branch based on the master branch, but I do not understand what I need to do to for Step 3 of the Contributing doc. @megies, would you mind helping me out with that? Or pointing me towards some more information on how to write a test for this.

Thank you,
Amanda
